### PR TITLE
Add 'date' as alias to 'day' for startOf() and endOf().

### PR DESCRIPTION
**Pull request by datyayu**
Patch for #2974.

### CODE REVIEW

File: src/lib/moment/start-end-of.js [hunk 1]
```diff
@@ -48,5 +49,11 @@ export function endOf (units) {
     if (units === undefined || units === 'millisecond') {
         return this;
     }
+
+    // 'date' is an alias for 'day', so it should be considered as such.
+    if (units === 'date') {
```
Comment (icambron) [inline src/lib/moment/start-end-of.js]: Can we instead just do

```js
if (units === 'date') {
  units = 'day';
}
```

**Comment by datyayu** [comment]
@icambron Sure, makes sense. Updated.

**Comment by icambron** [comment]
LGTM, @ichernev

**Comment by mj1856** [comment]
I don't understand why the if statement is needed at all in the `endOf` function.  Shouldn't it just be handled by the case statement in the `startOf` function?

For that matter, looking closer here, I don't understand why in the existing code there's an inline if for treating an `isoWeek` as a locale-based `week`, or why `endOf` is shortcutting for undefined and `millisecond` but `startOf` doesn't.

IMHO, all of the logic should be in the `startOf` function.  The `endOf` function should be consistent in going to the start of the next unit and then back one millisecond.

**Comment by datyayu** [comment]
Sorry for the delay, I've been busy lately.

The if statement sets the `units` to a valid value for the `add` method. Maybe the inline `isoWeek` handles the same case.

As for the `units === undefined` I think it isn't on `startOf` because `startOf` always returns `this` and ignores unhandled cases, but `endOf` returns by default a whole method chain, which is why it returns `this` on the `undefined` case, to avoid having to perform the operations with a wrong value. I'm not really sure about why the handling for the milliseconds..

**Comment by maggiepint** [comment]
I want to keep this pull request moving along, so:

I agree that undefined shortcutting is correct because of the return of the whole method chain.

Yes, date needs to be changed to day in endOf to make the add call valid. 

This actually opens up a bit of a rabbit hole. Ideally, by this logic, date would be an alias for day in the add/subtract functions. And that code path will lead you down to making date and alias for day in durations. I am not sure if this spirals out further, and it's late enough at night that I don't want to find out right now.

If we don't want to go make the day/date aliasing thing work everywhere, I think this PR is generally fine. I'm open to the opinions of others on whether we should carry the aliasing through. Programatically it makes sense to carry it all the way through, but from an API design standpoint it's a little wonky. Why would one add a date? One adds a day.

**Comment by ichernev** [comment]
We'll merge this, the code is not perfect, but it works :)

**Comment by ichernev** [comment]
Merged in 53934858a08d0ab461ceffc0c44c1a8fe8054eb5

**Pull request closed without merging**